odl/trafos/fourier.py

```diff
@@ -1645,6 +1647,211 @@ def inverse(self):
+    """
+    def __init__(self, shape, non_uniform_samples, domain, range):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        shape : tuple
+            The dimensions of the data whose non uniform FFT you want to
+            compute
+        non_uniform_samples : iterable
+            List of the fourier space positions where the coefficients are
+            computed. Each position must be in [-0.5; 0.5[
+        domain : `DiscreteLp`
+            Domain of the non uniform FFT or its adjoint
+        range : `DiscreteLp`
+            Range of the non uniform FFT or its adjoint
+        """
+        super(NonUniformFourierTransformBase, self).__init__(
+            domain=domain,
+            range=range,
+            linear=True,
+        )
+        self.shape = shape
+        self._check_samples(non_uniform_samples)
+        self.non_uniform_samples = non_uniform_samples
+        self.nfft = NFFT(N=shape, M=len(non_uniform_samples))
+        self.nfft.x = non_uniform_samples
+        self.nfft.precompute()
+        self.adjoint_class = None
+
+    def _check_samples(self, non_uniform_samples):
+        """Check that samples are correct
+
+        Parameters
+        ----------
+        non_uniform_samples : iterable
+            List of the fourier space positions where the coefficients are
+            computed
+
+        Raises
+        ------
+        ValueError :
+            If the non_uniform_samples is not an iterable, of 0 length or if
+            one of the locations is not normalized in [-0.5; 0.5[
+        """
+        if not isinstance(non_uniform_samples, Iterable):
+            raise TypeError('`non_uniform_samples` is not iterable.')
+
+        if len(non_uniform_samples) == 0:
+            raise ValueError('`non_uniform_samples` is empty')
+
+        n_dim = len(self.shape)
+        if not all(len(sample) == n_dim for sample in non_uniform_samples):
+            raise ValueError('One of the samples in `non_uniform_samples` does'
+              ' not have the right dimension')
+
+        if not all(
+                self._is_sample_valid(sample) for sample in non_uniform_samples
+            ):
+            raise ValueError('One of the samples in `non_uniform_samples` is'
+              ' not a float between -0.5 and 0.5')
+
+    def _is_sample_valid(self, sample):
+        """Check whether a single sample is correct.
+
+        Parameters
+        ----------
+        sample : iterable
+            The sample whose location you want to check
+
+        Returns
+        -------
+        bool : Whether the sample contains only floats in [-0.5; 0.5[
+        """
+        floats = all(isinstance(coord, float) for coord in sample)
+        if floats:
+            normalized = all(-0.5 <= coord < 0.5 for coord in sample)
+        else:
+            normalized = False
+        return floats & normalized
+
+    def _normalize(self, x):
+        """Normalize the result of the non uniform FFT.
```

Comment:
Please elaborate (on the row below) in what sense the normalization should be done.

Reply:
Do you mean how the normalization should be done or why?